sklearn/datasets/_arff_parser.py

```diff
@@ -206,7 +207,13 @@ def _io_to_generator(gzip_file):
         # liac-arff parser does not depend on NumPy and uses None to represent
         # missing values. To be consistent with the pandas parser, we replace
         # None with np.nan.
-        frame = pd.concat(dfs, ignore_index=True).fillna(value=np.nan)
+        frame = pd.concat(dfs, ignore_index=True)
+        pd_version = parse_version(pd.__version__).base_version
+        if parse_version(pd_version) < parse_version("2.2"):
+            frame = frame.fillna(value=np.nan)
+        else:
+            pd.set_option("future.no_silent_downcasting", True)
```

Comment:
Is this line really needed? If I read the warning, it seems like:
- you want the old behaviour: use `result.infer_objects(copy=False)`
- you want what will become the new behaviour: `pd.set_option('future.no_silent_downcasting', True)`

But you are doing both ...

Here is the warning for better context:
```
FutureWarning: Downcasting object dtype arrays on .fillna, .ffill, .bfill is
deprecated and will change in a future version. Call result.infer_objects(copy=False)
instead. To opt-in to the future behavior, set
`pd.set_option('future.no_silent_downcasting', True)`
```

Reply:
I tried and it looks like you need both indeed, I don't understand too much why for now ...

Reply:
You reach with the same conclusion :). Basically, if you start to use `downcasting=False` then you find out that the parameter is also deprecated so you end up with this code.

Reply:
Thanks for using the context manager. This is indeed safer to not have any side-effect.

Reply:
I looked a bit more what was going on. For the current release of pandas if for some columns, your dataframe has only missing values, `df.fillna(np.nan)` will change its dtype from `object` to `float64`.


```py
import numpy as np
import pandas as pd

df = pd.DataFrame({'col': [None, None, None]})

df.dtypes  # col has dtype object
df.fillna(np.nan).tdypes  # col has dtype float64
```

So it looks like indeed you need `pd.option_context('future.no_silent_downcasting', True)` to leave the dtype alone and still `.infer_objects` to get our old behaviour which was to have a float64 dtype for these columns.